relays/ethereum/src/ethereum_client.rs

```diff
@@ -150,6 +214,173 @@ async fn transaction_receipt(client: Client, hash: H256) -> (Client, Result<Rece
+	}
+
+	(client, Ok(HeaderId(number.low_u32(), hash)))
+}
+
+/// Returns true if Substrate header is known to Ethereum node.
+pub async fn substrate_header_known(
+	client: Client,
+	contract_address: Address,
+	id: SubstrateHeaderId,
+) -> (Client, Result<(SubstrateHeaderId, bool), Error>) {
+	let (encoded_call, call_decoder) = bridge_contract::functions::is_known_header::call(id.1);
+	let call_request = bail_on_arg_error!(
+		to_value(CallRequest {
+			to: Some(contract_address),
+			data: Some(encoded_call.into()),
+		})
+		.map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, call_result) =
+		bail_on_error!(call_rpc::<Bytes>(client, "eth_call", Params::Array(vec![call_request]),).await);
+	match call_decoder.decode(&call_result.0) {
+		Ok(is_known_block) => (client, Ok((id, is_known_block))),
+		Err(error) => (client, Err(Error::ResponseParseFailed(format!("{}", error)))),
+	}
+}
+
+/// Submits Substrate headers to Ethereum contract.
+pub async fn submit_substrate_headers(
+	client: Client,
+	params: EthereumSigningParams,
+	contract_address: Address,
+	headers: Vec<QueuedSubstrateHeader>,
+) -> (Client, Result<Vec<SubstrateHeaderId>, Error>) {
+	let (mut client, mut nonce) =
+		bail_on_error!(account_nonce(client, params.signer.address().as_fixed_bytes().into()).await);
+
+	let ids = headers.iter().map(|header| header.id()).collect();
+	for header in headers {
+		let (ret_client, _) = bail_on_error!(
+			submit_ethereum_transaction(
+				client,
+				&params,
+				Some(contract_address),
+				Some(nonce),
+				true, // we may need slightly more gas because other actors could change contract state
+				bridge_contract::functions::import_header::encode_input(header.extract().0.encode(),),
+			)
+			.await
+		);
+
+		nonce += 1.into();
+		client = ret_client;
+	}
+
+	(client, Ok(ids))
+}
+
+/// Deploy bridge contract.
+pub async fn deploy_bridge_contract(
+	client: Client,
+	params: &EthereumSigningParams,
+	contract_code: Vec<u8>,
+	initial_header: Vec<u8>,
+	initial_set_id: u64,
+	initial_authorities: Vec<u8>,
+) -> (Client, Result<(), Error>) {
+	submit_ethereum_transaction(
+		client,
+		params,
+		None,
+		None,
+		false,
+		bridge_contract::constructor(contract_code, initial_header, initial_set_id, initial_authorities),
+	)
+	.await
+}
+
+/// Submit ethereum transaction.
+async fn submit_ethereum_transaction(
+	client: Client,
+	params: &EthereumSigningParams,
+	contract_address: Option<Address>,
+	nonce: Option<U256>,
+	double_gas: bool,
+	encoded_call: Vec<u8>,
+) -> (Client, Result<(), Error>) {
+	let (client, nonce) = match nonce {
+		Some(nonce) => (client, nonce),
+		None => bail_on_error!(account_nonce(client, params.signer.address().as_fixed_bytes().into()).await),
+	};
+	let (client, gas) = bail_on_error!(
+		estimate_gas(
+			client,
+			CallRequest {
+				to: contract_address,
+				data: Some(encoded_call.clone().into()),
+				..Default::default()
```

Comment:
This is more of a dumb Rust question, but why's this here? `CallRequest` has two fields which are already set, so it seems unnecessary to me - but I might be missing something